psl-core/src/main/java/org/linqs/psl/util/VizDataCollection.java

```diff
@@ -0,0 +1,233 @@
+            double violation = 0.0;
+            boolean weightFlag = false;
+            Iterable<GroundRule> groundedRuleList = groundRuleStore.getGroundRules(rule);
+            for (GroundRule groundRule : groundedRuleList) {
+                if (vizData.violatedGroundRulesList.contains(groundRule)) {
+                    //There can't be weighted violated rules so we can make an assumption here
+                    UnweightedGroundRule unweightedGroundRule = (UnweightedGroundRule)groundRule;
+                    violation = unweightedGroundRule.getInfeasibility();
+                    moduleElement.put("Violated Rule", groundRule.baseToString());
+                    moduleElement.put("Parent Rule", rule.getName());
+                    // moduleElement.put("Weighted", weightFlag);
+                    moduleElement.put("Violation", violation);
+                    vizData.violatedGroundRulesArray.put(moduleElement);
+                }
+            }
+        }
+    }
+
+    public static void ruleMapInsertElement(AbstractLogicalRule parentRule, GroundRule groundRule,
+                            Map<Variable, Integer> variableMap,  Constant[] constantsList) {
+      //Adds a groundAtom element to RuleMap
+      //Why are some groundRules null? Perhaps a thread thing??
+      ArrayList<Integer> atomHashList = new ArrayList<>();
+      if (groundRule != null) {
+        HashSet<Atom> atomSet = new HashSet<>(groundRule.getAtoms());
+        int atomCount = 0;
+        HashMap<String,String> atomMap = new HashMap<>();
+        for (Atom a : atomSet) {
+          atomHashList.add(System.identityHashCode(a));
+          vizData.groundAtoms.put(Integer.toString(System.identityHashCode(a)), a.toString());
+          atomCount++;
+        }
+      }
+
+      //Adds a rule element to RuleMap
+      JSONObject rulesElement = new JSONObject();
+      String ruleStringID = Integer.toString(System.identityHashCode(parentRule));
+      JSONObject rulesElementItem = new JSONObject();
+      rulesElementItem.put("text", parentRule);
+      vizData.rules.put(ruleStringID, rulesElementItem);
+
+      //Adds a groundRule element to RuleMap
+      HashMap<String, String> varConstMap = new HashMap<>();
+      for (Map.Entry<Variable, Integer> entry : variableMap.entrySet()) {
+          varConstMap.put(entry.getKey().toString(), constantsList[entry.getValue()].rawToString());
+      }
+      JSONObject groundRulesElement = new JSONObject();
+      groundRulesElement.put("ruleID", Integer.parseInt(ruleStringID));
+      JSONObject constants = new JSONObject();
+      for (Map.Entry varConstElement : varConstMap.entrySet()) {
+        String key = (String)varConstElement.getKey();
+        String val = (String)varConstElement.getValue();
+        constants.put(key,val);
+      }
+      groundRulesElement.put("constants", constants);
+      groundRulesElement.put("groundAtoms", atomHashList);
+      //We dont get any null groundRules here???
```

Comment:
You do, but `System.identityHashCode(null) == 0`.

Reply:
Resolved.